File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.cpp
class Solution {
public:
    int minSetSize(vector<int>& arr) {
        unordered_map<int, int> freq;
        for (auto &num : arr) {
            freq[num]++;
        }
        vector<int> freq_list;
        for (auto &it : freq) {
            freq_list.push_back(it.second);
        }
        sort(freq_list.begin(), freq_list.end());
        int target = arr.size() / 2;
        int removed = 0;
        int count = 0;
        while (removed < target) {
            removed += freq_list.back();
            freq_list.pop_back();
            count++;
        }
        return count;
    }
};

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.py
from collections import defaultdict
class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        freq = defaultdict(int)
        for num in arr:
            freq[num] += 1
        freq_list = list(freq.values())
        freq_list.sort()
        target = len(arr) // 2
        removed = 0
        count = 0
        while removed < target:
            removed += freq_list.pop()
            count += 1
        return count
